FEAT Add easier querying for supported input/output types

paulinek13 opened the pull request:
<!--- Please add one of the following as a prefix to the pull request title: -->
<!--- DOC for documentation changes -->
<!--- MAINT for maintenance changes, e.g., build pipeline fixes -->
<!--- FIX for bug fixes -->
<!--- TEST for adding tests -->
<!--- FEAT for new features and enhancements (which implies that tests + doc changes are included) -->
<!--- Additionally, if your PR is not yet ready for review, create it as a "Draft" PR and prefix [DRAFT] -->

## Description
<!--- Provide a general summary of your changes. -->
<!--- Mention related issues, pull requests, or discussions with #<issue/PR/discussion ID>. -->
<!--- Tag people for whom this PR may be of interest using @<username>. -->

<!--- If you are considering making a contribution please open an issue first. -->
<!--- This can help in identifying if the contribution fits into the plans for PyRIT. -->
<!--- Maintainers may be aware of obstacles that aren't obvious, or clarify requirements, and thereby save you time. -->

This PR aims to resolve #543 by adding a simpler way to check which input/output types a converter supports.

**Implementation Tasks**

- [x] Add `supported_input_types` property to `PromptConverter` class
- [x] Add `output_supported` method, mirroring `input_supported`
- [x] Add `supported_output_types` property to `PromptConverter` class
- [x] ~Update `PromptDataType` literal with all supported types~
- [x] Add unit tests for the new properties
- [x] ~Fix failing `tests/unit/test_literals.py::test_prompt_data_type` test due to `PromptDataType` changes~

This is a ~draft~ PR seeking early feedback on the approach.

<!--- ## Tests and Documentation -->

<!--- Contributions require tests and documentation (if applicable). -->
<!--- Include a description of tests and documentation updated (if applicable) -->

<!--- JupyText helps us see regressions in APIs or in our documentation by executing all code samples -->
<!--- Include how you/if ran JupyText here -->
<!--- This is described at: https://github.com/Azure/PyRIT/tree/main/doc  -->

rlundeen2 commented on pyrit/models/literals.py:
```diff
@@ -3,7 +3,8 @@
 from typing import Literal
 
-PromptDataType = Literal["text", "image_path", "audio_path", "video_path", "url", "error"]
+PromptDataType = Literal["audio", "audio_path", "image", "image_path", "non-text",
```
My initial thought is that don't want to update this; but is there a reason we'd want the new types?

paulinek13 replied:
I found some parameters that are being used with `input_supported` but aren't defined in `PromptDataType`, and so I added the missing ones to match what's in the code. That's my reasoning behind this change

rlundeen2 replied:
In theory that should cause a mypy error in actual - let's change it back here because the code doesn't know how to handle "image" type so things will break. As we find code that's using it correctly (which my guess is mocked types) we should probably change that to a supported `PromptDataType` though

rlundeen2 commented on pyrit/prompt_converter/prompt_converter.py:
```diff
@@ -100,3 +112,23 @@ def get_identifier(self):
+
+        Returns:
+            list[PromptDataType]: A list of supported input types.
+        """
```
Can you write basic unit tests for these methods?

paulinek13 replied:
Yes, sure!

rlundeen2 reviewed (approved):
I like this!

There are two changes I'd like but once tests are added and PromptDataTypes are either changed back (or we talk about it) I think ready to merge

paulinek13 commented on pyrit/prompt_converter/prompt_converter.py:
```diff
@@ -100,3 +112,23 @@ def get_identifier(self):
+        Returns:
+            list[PromptDataType]: A list of supported output types.
+        """
+        return [data_type for data_type in PromptDataType.__args__ if self.output_supported(data_type)]
```
The current build fails:
`pyrit/prompt_converter/prompt_converter.py:134: error: "<typing special form>" has no attribute "__args__"  [attr-defined]`

I'll fix that

paulinek13 commented:
I've made the following changes:

1. Reverted `PromptDataType` to its original state
2. Added tests for `supported_input_types` and `supported_output_types` properties for converters

I tested it locally:

- [x] `pytest tests/unit` completes without errors
- [x] `pre-commit run --all-files` runs successfully

I also noticed that several converters are missing from the jupyter-book docs, even though there are documented in the code: `AsciiSmugglerConverter`, `CharSwapGenerator`, `ColloquialWordswapConverter`, `DiacriticConverter`.
